src/main/java/de/tum/in/www1/artemis/service/scheduled/ProgrammingExerciseScheduleService.java

```diff
@@ -191,6 +195,20 @@ private void scheduleCourseExercise(ProgrammingExercise exercise) {
         }
 
         // For any course exercise that needsToBeScheduled (dueDate and/or manual assessment)
+
+        // For any course exercise with a valid release date
+        if (exercise.getReleaseDate() != null && ZonedDateTime.now().isBefore(exercise.getReleaseDate())) {
+            var scheduledRunnable = Set.of(new Tuple<>(exercise.getReleaseDate().minusSeconds(Constants.SECONDS_BEFORE_RELEASE_DATE_FOR_COMBINING_TEMPLATE_COMMITS),
+                    combineTemplateCommitsForExercise(exercise)));
+            scheduleService.scheduleTask(exercise, ExerciseLifecycle.RELEASE, scheduledRunnable);
+            log.debug("Scheduled combining template commits before release date for Programming Exercise \"" + exercise.getTitle() + "\" (#" + exercise.getId() + ") for "
```

Comment:
We should let the logger format the Strings. This is more efficient, as e.g. in production, the debug level is inactive. This can be done with `.. {} ..`, similar to `String.format`.

Reply:
(Looks like we need to change that in many places, though)

Reply:
Could be done in a separate PR.